token/cli/src/config.rs

```diff
@@ -276,42 +322,52 @@ impl<'a> Config<'a> {
         override_name: &str,
         wallet_manager: &mut Option<Arc<RemoteWalletManager>>,
         token: Option<Pubkey>,
-    ) -> Pubkey {
-        if let Some(address) = pubkey_of_signer(arg_matches, override_name, wallet_manager).unwrap()
+    ) -> Result<Pubkey, Error> {
+        if let Some(address) = pubkey_of_signer(arg_matches, override_name, wallet_manager).ok().flatten()
         {
-            return address;
+            return Ok(address);
         }
 
         let token = token.unwrap();
         let program_id = self.get_mint_info(&token, None).await.unwrap().program_id;
-        self.associated_token_address_for_token_and_program(&token, &program_id)
+        let owner = self.maybe_pubkey(arg_matches, "owner", wallet_manager);
+        self.associated_token_address_for_token_and_program(&token, owner.as_ref(), &program_id).await
     }
 
-    pub(crate) fn associated_token_address_for_token_and_program(
+    pub(crate) async fn associated_token_address_for_token_and_program(
```

Comment:
why is this async?

Reply:
> why is this async?

I never wrote async Rust code before.  I thought that I had to make it async in order to be able to use the '?' operator.  If I recall if I didn't make it async I got really gnarly errors.

Reply:
Ah I see the confusion.  The `?` is meant for functions that return `Result<T, E>` or `Option<T>`, and allow you to short-circuit if the value is is an error or `None`, respectively.  It's extremely useful!  https://doc.rust-lang.org/rust-by-example/std/result/question_mark.html has more info.  Either way, I think you should be able remove the `async`

Reply:
Yeah so this is where I get really confused.

For example after changing all the code to the default_signer()?.pubkey() version instead of having a default_signer_pubkey() function ...

I end up with an error like this:

```
error[E0282]: type annotations needed
   --> token/cli/src/config.rs:331:36
    |
331 |         let owner = owner.map(|pk| Ok(pk.clone())).unwrap_or_else(|| Ok(self.default_signer()?.pubkey()))?;
    |                                    ^^ cannot infer type for type parameter `E` declared on the enum `Result`
    |
help: consider specifying the type arguments in the method call
    |
331 |         let owner = owner.map::<U, F>(|pk| Ok(pk.clone())).unwrap_or_else(|| Ok(self.default_signer()?.pubkey()))?;
    |                              ++++++++

error[E0283]: type annotations needed
   --> token/cli/src/config.rs:331:106
    |
331 |         let owner = owner.map(|pk| Ok(pk.clone())).unwrap_or_else(|| Ok(self.default_signer()?.pubkey()))?;
    |                                                                                                          ^ cannot infer type for type parameter `E`
    |
    = note: multiple `impl`s satisfying `Box<dyn StdError + Send + Sync>: ~const From<_>` found in the following crates: `anyhow`, `core`, `std`:
            - impl From<anyhow::Error> for Box<(dyn StdError + Send + Sync + 'static)>;
            - impl From<std::string::String> for Box<(dyn StdError + Send + Sync + 'static)>;
            - impl<'a, 'b> From<Cow<'b, str>> for Box<(dyn StdError + Send + Sync + 'a)>;
            - impl<'a, E> From<E> for Box<(dyn StdError + Send + Sync + 'a)>
              where E: 'a, E: StdError, E: Send, E: Sync;
            and 2 more
    = note: required because of the requirements on the impl of `FromResidual<Result<std::convert::Infallible, _>>` for `Result<Pubkey, Box<dyn StdError + Send + Sync>>`
help: consider specifying the type arguments in the method call
    |
331 |         let owner = owner.map::<U, F>(|pk| Ok(pk.clone())).unwrap_or_else(|| Ok(self.default_signer()?.pubkey()))?;
    |                              ++++++++

error[E0308]: mismatched types
   --> token/cli/src/config.rs:358:9
    |
352 |     ) -> Result<Pubkey, Error> {
    |          --------------------- expected `Result<Pubkey, Box<(dyn StdError + Send + Sync + 'static)>>` because of return type
...
358 |         self.default_signer()?.pubkey()
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected enum `Result`, found struct `Pubkey`
    |
    = note: expected enum `Result<Pubkey, Box<(dyn StdError + Send + Sync + 'static)>>`
             found struct `Pubkey`
help: try wrapping the expression in `output::_::_serde::__private::Ok`
    |
358 |         output::_::_serde::__private::Ok(self.default_signer()?.pubkey())
    |         +++++++++++++++++++++++++++++++++                               +

Some errors have detailed explanations: E0282, E0283, E0308.
For more information about an error, try `rustc --explain E0282`.
error: could not compile `spl-token-cli` due to 3 previous errors
```

Basically what is happening here is that I'm trying to use ? to handle the error when calling self.default_signer()? but since default_signer() is not async but the calling function is, I get type mismatches.

Reply:
Well I can fix the above by using more verbose syntax:

```
        let owner = {
            if let Some(pubkey) = owner {
                pubkey.clone()
            }
            else {
                self.default_signer()?.pubkey()
            }
        };
```

Reply:
OK anyway I was totally wrong, removing async from associated_token_address_for_token_and_program "just worked" so I guess it was part of an incorrect fix that I made when I was trying to get things to compile.